src/Components/Components/src/Forms/ValidationMessageStore.cs

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+
+namespace Microsoft.AspNetCore.Components.Forms
+{
+    /// <summary>
+    /// Holds validation messages for an <see cref="EditContext"/>.
+    /// </summary>
+    public class ValidationMessageStore
+    {
+        private readonly EditContext _editContext;
+        private readonly Dictionary<FieldIdentifier, List<string>> _messages = new Dictionary<FieldIdentifier, List<string>>();
+
+        /// <summary>
+        /// Creates an instance of <see cref="ValidationMessageStore"/>.
+        /// </summary>
+        /// <param name="editContext">The <see cref="EditContext"/> with which this store should be associated.</param>
+        public ValidationMessageStore(EditContext editContext)
+        {
+            _editContext = editContext ?? throw new ArgumentNullException(nameof(editContext));
+        }
+
+        /// <summary>
+        /// Adds a validation message for the specified field.
+        /// </summary>
+        /// <param name="fieldIdentifier">The identifier for the field.</param>
+        /// <param name="message">The validation message.</param>
+        public void Add(FieldIdentifier fieldIdentifier, string message)
+            => GetOrCreateMessagesListForField(fieldIdentifier).Add(message);
+
+        /// <summary>
+        /// Adds the messages from the specified collection for the specified field.
+        /// </summary>
+        /// <param name="fieldIdentifier">The identifier for the field.</param>
+        /// <param name="messages">The validation messages to be added.</param>
+        public void AddRange(FieldIdentifier fieldIdentifier, IEnumerable<string> messages)
+            => GetOrCreateMessagesListForField(fieldIdentifier).AddRange(messages);
+
+        /// <summary>
+        /// Gets the validation messages within this <see cref="ValidationMessageStore"/> for the specified field.
+        ///
+        /// To get the validation messages across all validation message stores, use <see cref="EditContext.GetValidationMessages(FieldIdentifier)"/> instead
+        /// </summary>
+        /// <param name="fieldIdentifier">The identifier for the field.</param>
+        /// <returns>The validation messages for the specified field within this <see cref="ValidationMessageStore"/>.</returns>
+        public IEnumerable<string> this[FieldIdentifier fieldIdentifier]
```

Comment:
Another thing we could consider for these APIs is using `IReadOnlyList<>`. It's the non-allocating big brother of `IEnumerable<>`

Reply:
We could, but it doesn't help developers much in this case. There isn't much use case in reading back the messages from a single store. Normally you only want to read back all the messages for a given field across all stores in the `EditContext`, which has to be `IEnumerable<string>` because it dynamically combines messages from all stores (and we don't want to build lists on each evaluation).